parquet-column/src/main/java/org/apache/parquet/internal/column/columnindex/ColumnIndexBuilder.java

```diff
@@ -368,6 +369,102 @@ public <T extends Comparable<T>> PrimitiveIterator.OfInt visit(NotIn<T> notIn) {
       return IndexIterator.all(getPageCount());
     }
 
+    @Override
+    public <T extends Comparable<T>> PrimitiveIterator.OfInt visit(Contains<T> contains) {
+      return contains.filter(
+          this,
+          // For Contains.And, compute the intersection of the int ranges
+          (l, r) -> new PrimitiveIterator.OfInt() {
+            private Integer next = null;
+
+            @Override
+            public int nextInt() {
+              int result = next;
+              next = null;
+              return result;
+            }
```

Comment:
makes sense! I had put some of the next-loading logic into `hasNext()` because as far as I could tell, this iterator would be used via the `forEachRemaining()` method, which IIRC does sequentially call `hasNext()/next()`... but it does make this a bit unstable 😅 I can rewrite it and move into static methods in `IndexIterator`.